src/WindowTitleBar/CloneButton.tsx

```diff
@@ -0,0 +1,173 @@
+		if(err) {
+			await FSBL.Clients.WindowClient.createTabCollection({windowNames: [finsembleWindow.name, name], promotedTab:name});
+			return;
+		}
+
+		if(data) {
+			await FSBL.Clients.WindowClient.addToTabCollection({
+				tabCollectionId: data.id,
+				windowNames: [name],
+				promotedTab: name
+			});
+		}
+	}
+
+	/**
+	 * Too slow. Linker does not change.
+	 */
+	const setComponentState = (fsblWindow: FinsembleWindow, componentState: Object) => {
+		const fields:any[] = [];
+		Object.keys(componentState).forEach((key) => {
+			fields.push({field:key, value:componentState[key]})
+		})
+
+		fsblWindow.setComponentState({fields})
+	}
+
+	/**
+	 * Originally used to set component state before spawn but set it immediately after spawn - potential race condition.
+	 * @param windowName
+	 * @param componentState
+	 */
+	const prepareComponentState = async (windowName:string, componentState:any) => {
+		const topic = "finsemble.workspace.cache"
+		await FSBL.Clients.StorageClient.save({
+			topic,
+			key: `activeWorkspace${windowName}${windowName}`,
+			value: componentState
+		})
+	}
+
+	/**
+	 * Unused. But can be used to clone a window other than the current window
+	 * (potentially not correct url if remote clone)
+	 * Add to tab group would need to be updated for remote clone.
+	 * @param windowName
+	 */
+	const getWindow = async (windowName) => {
+		const {wrap} = await FSBL.FinsembleWindow.getInstance({
+			windowName
+		})
+
+		return wrap;
+	}
+
+
+	/**
+	 *
+	 * @param fsblWindow
+	 */
+	const getComponentData = async (fsblWindow:FinsembleWindow): Promise<{component: string, spawnParams: SpawnParams, componentState:any}> => {
+		const {windowOptions} = fsblWindow;
+
+		// Component Type
+		const component = windowOptions.componentType;
+
+		// Get the URL
+		// Potential bug here? Is this the current url?
+		const url = windowOptions.url;
+
+		// Window dimensions and position
+		const {data:currBounds} = await finsembleWindow.getBounds();
+
+		const bounds:any = {};
+		if (currBounds) {
+			bounds.left = currBounds.left
+			bounds.top = currBounds.top
+			bounds.width = currBounds.width
+			bounds.height = currBounds.height
+		}
+
+		// component state
+		const componentState =  await getAsyncComponentState(fsblWindow);
+
+		return {
+			component,
+			componentState,
+			spawnParams:{
+				url,
+				...bounds,
+				name:generateWindowName(component)
+			}};
+	}
+
+	/**
+	 * Generates a window name for the clone. Used to set components state as quickly as possible
+	 * @param componentType
+	 */
+	const generateWindowName = (componentType):string => {
+		//Note: This will only work in localhost or a secure environment.
+		return `${componentType}-clone-${crypto.randomUUID()}`;
+	}
+
+	/**
+	 * simplified getComponentState and returning a promise
+	 * @param win
+	 */
+	const getAsyncComponentState = (fsblWindow:FinsembleWindow) => {
```

Comment:
When moving this into service, maybe just go fix these functions to return promises